### ISSUE TITLE
Clarification on using --train-num-samples (lower value) without --dataset-resampled

### ISSUE THREAD

**Issue by fadamsyah** [mlfoundations/open_clip#993]
> For larger datasets (eg Laion2B), we recommend setting --train-num-samples to a lower value than the full epoch, for example --train-num-samples 135646078 to 1/16 of an epoch in conjunction with --dataset-resampled to do sampling with replacement. This allows having frequent checkpoints to evaluate more often.

What happens if I use --train-num-samples with a lower value (e.g., 1/16 of an epoch) but do not use --dataset-resampled? Will the training still work correctly, and are there any downsides to this approach? My main aim is to save more frequent checkpoints.

Thank you!

**Comment by rom1504** [comment]
It will use only a subset of the dataset, so probably not what you want

**Comment by fadamsyah** [comment]
Hi @rom1504,

Thank you for the clarification! Could you please share if there’s a specific reason why --dataset-resampled performs sampling with replacement instead of without replacement?

**Comment by rwightman** [comment]
@fadamsyah as per what @rom1504 says, it basically ends each 'epoch' at that specified sample count so each epoch won't be a full one. But since each epoch begins with a different shard shuffle I don't see it being an issue over the length of training, I doubt you'd see a difference between say 64 epochs and train-samples being 1/2 the actual vs 32 with full.

I feel it would be an okay way to get more frequent checkpoints and be able to resume with some consistency. Each epoch shuffle should be same for same seed / world details. The only difference between this and resampled is that you won't have likely repeat samples in same epoch...

**Comment by rwightman** [comment]
it's not practical to sample without replacement across many dataloader workers and distributed processes, it'd be rather challenging and impractical to implement

**Comment by rwightman** [comment]
I guess one difference is over the duration of training, the count of each sample being seen might differ more with this approach, e.g. for 64 epochs with 1/2 sample count, some samples might be seen 29,30,31,32,33,34 etc times with some dist.

**Comment by fadamsyah** [comment]
Thanks so much for the clarification @rwightman ! I appreciate the thoughtful explanations. I’ll consider giving it a try
